### PR TITLE
[Core][1/N] Support PP PyNCCL Groups

### PR DESCRIPTION
**Pull request by andoorve**
This PR adds supports multiple PP groups. Will help https://github.com/vllm-project/vllm/pull/4412 support CUDAGraph.

### CODE REVIEW

**Comment by andoorve** [comment]
@youkaichao @simon-mo Can you take a quick look at this PR?

File: tests/distributed/test_pynccl.py [hunk 1]
```diff
@@ -151,6 +152,68 @@ def test_pynccl_with_cudagraph():
     distributed_run(worker_fn_with_cudagraph, 2)
 
 
+@worker_fn_wrapper
+def pp_worker_fn():
```
Comment (youkaichao) [inline tests/distributed/test_pynccl.py:156]: name it as `send_recv` worker? it is not testing pp though.

File: vllm/distributed/device_communicators/pynccl.py [hunk 2]
```diff
@@ -126,6 +126,34 @@ def all_reduce(self,
...
+    def send(self, tensor: torch.Tensor, stream=None):
+        if self.disabled:
+            return
+        assert tensor.device == self.device, (
+            f"this nccl communicator is created to work on {self.device}, "
+            f"but the input tensor is on {tensor.device}")
+        if stream is None:
+            stream = self.stream
+        # Hardcoded to send to the next rank for now for simplicity.
+        dst = (self.rank + 1) % self.world_size
```
Comment (youkaichao) [inline vllm/distributed/device_communicators/pynccl.py:138]: why hardcode here?

File: vllm/distributed/device_communicators/pynccl.py [hunk 3]
```diff
@@ -126,6 +126,34 @@ def all_reduce(self,
...
+    def recv(self, tensor: torch.Tensor, stream=None):
+        if self.disabled:
+            return
+        assert tensor.device == self.device, (
+            f"this nccl communicator is created to work on {self.device}, "
+            f"but the input tensor is on {tensor.device}")
+        if stream is None:
+            stream = self.stream
+        # Hardcoded to receive from the previous rank for now for simplicity.
+        src = (self.rank - 1) % self.world_size
```
Comment (youkaichao) [inline vllm/distributed/device_communicators/pynccl.py:152]: same as above

File: vllm/distributed/parallel_state.py [hunk 4]
```diff
@@ -195,17 +202,25 @@ def initialize_model_parallel(
...
         if rank in ranks:
             _PP_DEVICE_GROUP = group
+            _PP_CPU_GROUP = cpu_group
             _PP_GLOBAL_RANKS = ranks
 
+    _PP_PYNCCL_COMMUNICATOR = PyNcclCommunicator(
+        group=_PP_CPU_GROUP,
+        device=_LOCAL_RANK,
+    )
+
```
Comment (youkaichao) [inline vllm/distributed/parallel_state.py:223]: creating communicators will add memory cost. please add a check here, if pp size == 1, skip the construction of `PyNcclCommunicator`.

File: tests/distributed/test_pynccl.py (hunk 1 above)
Comment (andoorve) [reply tests/distributed/test_pynccl.py:156]: Ok will change - this is following convention from here:

https://github.com/andoorve/vllm/blob/8824d36656e52507f341904df23c9475aaf8c0e7/tests/distributed/test_pynccl.py#L72

File: vllm/distributed/device_communicators/pynccl.py (hunk 2 above)
Comment (andoorve) [reply vllm/distributed/device_communicators/pynccl.py:138]: It's hardcoded as we only need to send to the next and prev rank for now so it's simple. We can change this if necessary though

File: tests/distributed/test_pynccl.py (hunk 1 above)
Comment (youkaichao) [reply tests/distributed/test_pynccl.py:156]: you can also change that into allreduce worker.

File: vllm/distributed/device_communicators/pynccl.py (hunk 2 above)
Comment (youkaichao) [reply vllm/distributed/device_communicators/pynccl.py:138]: please don't hardcode here. just expose the general form of `send/recv`. You can add default arg.

Comment (andoorve) [reply vllm/distributed/device_communicators/pynccl.py]: Changed it to expose general form with default arg `None`, which will make it next and prev rank respectively.

Comment (andoorve) [reply vllm/distributed/device_communicators/pynccl.py]: Let me know if that's sufficient

File: vllm/distributed/parallel_state.py (hunk 4 above)
Comment (andoorve) [reply vllm/distributed/parallel_state.py]: Changed both PP and TP.

File: tests/distributed/test_pynccl.py (hunk 1 above)
Comment (andoorve) [reply tests/distributed/test_pynccl.py]: Done

**Review by youkaichao — APPROVED** [review]
looks good to me overall. thanks! let's see if it passes the tests.

**Comment by andoorve** [comment]
Hey @youkaichao looks like it passed, can this be merged?

File: vllm/distributed/parallel_state.py (hunk 4 above)
Comment (SolitaryThinker) [reply vllm/distributed/parallel_state.py]: @youkaichao Do you know how much extra memory the communicators would be adding? It is a fixed cost or scales with group size? 

We are running into some OOM with cudagraph enabled that shouldn't be happening. Any pointers would be greatly appreciated!

**Pull request merged**